web-ui/src/streaming/builder/hooks/useAutoLayout.ts

```diff
@@ -0,0 +1,229 @@
+// Computes the layout of the nodes in the graph.
+
+import assert from 'assert'
+import { useCallback, useEffect } from 'react'
+import { Node, ReactFlowState, useStore, useReactFlow, Edge, Instance, getConnectedEdges } from 'reactflow'
+import { removePrefix } from 'src/utils'
+
+// How much spacing we put after every input/output node
+const VERTICAL_SPACING = 20
+const HORIZONTAL_SPACING = 180
+
+const nodeCountSelector = (state: ReactFlowState) => state.nodeInternals.size + state.edges.length
+const nodesInitializedSelector = (state: ReactFlowState) =>
+  Array.from(state.nodeInternals.values()).every(node => node.width && node.height)
+
+// We want to sort input nodes so that if we connect them to a program node, the
+// edges will not overlap each other.
+//
+// We do this by: Extract and sort connectors that have an edge Merge items that
+// don't have an edge yet with sorted connectors
+//
+// Also note that this approach won't work for positioning of the output node
+// because they can have multiple edges.
+function sortInputConnectors(nodes: Node[], edges: Edge[], tableOrder: Map<string, number>) {
+  // Find all nodes that have an edge attached to them
+  const nodesWithAnEdge = nodes.filter(node => {
+    return edges.some(edge => edge.source === node.id)
+  })
+
+  // Find the nodes that don't have an edge and remember the original index in
+  // `nodes` for them
+  const nodesAndIndexWithoutAnEdge: { [index: number]: Node } = nodes
+    .map((node, index) => {
+      return {
+        node,
+        index
+      }
+    })
+    .filter(element => {
+      return edges.every(edge => edge.source !== element.node.id)
+    })
+    .reduce((obj, item) => ({ ...obj, [item.index]: item.node }), {})
+
+  nodesWithAnEdge.sort((a, b) => {
+    const aEdges = getConnectedEdges([a], edges)
+    const bEdges = getConnectedEdges([b], edges)
+    assert(aEdges.length === 1)
+    assert(bEdges.length === 1)
+    assert(aEdges[0].targetHandle && bEdges[0].targetHandle)
+
+    const aTable = removePrefix(aEdges[0].targetHandle, 'table-')
+    const bTable = removePrefix(bEdges[0].targetHandle, 'table-')
+    assert(aTable && bTable)
+
+    const aIndex = tableOrder.get(aTable)
+    const bIndex = tableOrder.get(bTable)
+    assert(aIndex !== undefined && bIndex !== undefined)
+
+    return aIndex - bIndex
+  })
+
+  // Merge the two arrays
+  const result = []
+  let sortedIndex = 0
+  for (let i = 0; i < nodes.length; i++) {
+    if (!nodesAndIndexWithoutAnEdge[i]) {
+      result.push(nodesWithAnEdge[sortedIndex])
+      sortedIndex++
+    } else {
+      result.push(nodes[i])
+    }
+  }
+
+  return result
+}
+
+// Positions the nodes in a fixed layout. The layout is:
+//
+// - Input nodes are on the left
+// - Output nodes are on the right
+// - Program node is in the middle
+// - The Input and Output nodes are evenly spaced vertically.
```

Comment:
could a table be used for that?

Reply:
this is a good idea I will try it in the future as it involves quite a bit of a bit of refactoring. it it works will simplify the layout code